### PR TITLE
Fixed course variants bug. Added exceptions

### PR DESCRIPTION
**Pull request by MartinBentley**
### Context
Fix a bug when selecting a course variant.
### Changes proposed in this pull request
Fix bug and throw error rather than return null
### Guidance to review
The lambda query needs to select the course that contains a child variant with the required ucas code.
This was found to be incorrect and is now fixed.
Also exceptions now thrown when a variant or course details object is null
There are now unit tests for this controller. This required that the manageApi wrapper class has an interface so that it can be mocked.

### CODE REVIEW

**Comment by benilovj** [comment]
Is there a unit test that would have caught this issue?

File: src/ui/Controllers/CourseController.cs [hunk 1]
```diff
@@ -29,10 +29,19 @@ public CourseController(ManageApi manageApi)
...
+            if (courseDetail == null)
+            {
+                throw new Exception("Unexpected error");
+            }
 
             var variant = courseDetail.Variants.FirstOrDefault(v => v.UcasCode == ucasCode);
-            if (variant == null) return null;
+            if (variant == null)
+            {
+                throw new Exception("Unexpected error");
```
[the remark this answers is not in the document]
Comment (defong) [reply src/ui/Controllers/CourseController.cs]: That be good as there is two `throw new Exception("Unexpected error");` getting throwned

File: ManageCoursesUi.Tests/Helpers/TestHelper.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using GovUk.Education.ManageCourses.ApiClient;
+
+namespace ManageCoursesUi.Tests.Helpers
+{
+    /// <summary>
+    /// Defines the type of data that needs to be generated by the helper class
+    /// </summary>
+    public enum EnumDataType
```
Comment (timabell) [inline ManageCoursesUi.Tests/Helpers/TestHelper.cs]: One class/thing per file please

Minor.

File: ManageCoursesUi.Tests/CourseControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+            try
+            {
+                await controller.Variants(TestHelper.InstitutionCode, TestHelper.AccreditedProviderId, TestHelper.TargetedUcasCode);
+            }
+            catch (Exception e)
+            {
+                exceptionCalled = (e.Message == "Unexpected error: course should not be null");
+            }            
+            Assert.IsTrue(exceptionCalled);
```
Comment (timabell) [inline ManageCoursesUi.Tests/CourseControllerTests.cs]: Use fluent assertions. Much cleaner, better failure info. https://stackoverflow.com/a/35006901/10245

Reject.

File: src/ui/Controllers/CourseController.cs [hunk 4]
```diff
@@ -14,25 +14,36 @@ namespace GovUk.Education.ManageCourses.Ui.Controllers
...
                 ? course.ProviderCourses.First(c => String.IsNullOrEmpty(c.AccreditingProviderId))
                 : course.ProviderCourses
                     .First(c => c.AccreditingProviderId.Equals(accreditingProviderId, StringComparison.InvariantCultureIgnoreCase));
-            var courseDetail = providerCourse.CourseDetails.First(x => x.Variants.Select(v => v.UcasCode == ucasCode).Any());
+
+            var courseDetail = providerCourse.CourseDetails.FirstOrDefault(c => c.Variants.Any(v => v.UcasCode == ucasCode));
+
+            if (courseDetail == null)
+            {
+                throw new Exception("Unexpected error: course should not be null");
```
Comment (timabell) [inline src/ui/Controllers/CourseController.cs]: suggestion:

    throw new InvalidOperationException($"Course variant with ucas code '{ucasCode}' not found");

https://stackoverflow.com/questions/1453200/exception-for-missing-data

Minor.

File: src/ui/Controllers/CourseController.cs [hunk 5]
```diff
@@ -14,25 +14,36 @@ namespace GovUk.Education.ManageCourses.Ui.Controllers
...
+
             var course = await _manageApi.GetCoursesByOrganisation(instCode);
 
             var providerCourse = "self".Equals(accreditingProviderId, StringComparison.InvariantCultureIgnoreCase)
                 ? course.ProviderCourses.First(c => String.IsNullOrEmpty(c.AccreditingProviderId))
                 : course.ProviderCourses
                     .First(c => c.AccreditingProviderId.Equals(accreditingProviderId, StringComparison.InvariantCultureIgnoreCase));
-            var courseDetail = providerCourse.CourseDetails.First(x => x.Variants.Select(v => v.UcasCode == ucasCode).Any());
+
+            var courseDetail = providerCourse.CourseDetails.FirstOrDefault(c => c.Variants.Any(v => v.UcasCode == ucasCode));
```
Comment (timabell) [inline src/ui/Controllers/CourseController.cs]: Use `SingleOrDefault` if you only expect one.

https://enterprisecraftsmanship.com/2015/09/15/fail-fast-principle/

Reject.

File: src/ui/Controllers/CourseController.cs [hunk 6]
```diff
@@ -89,5 +100,12 @@ public CourseController(ManageApi manageApi)
 
             return View(viewModel);
         }
+
+        private void Validate(string instCode, string accreditingProviderId, string ucasCode)
+        {
+            if (string.IsNullOrEmpty(instCode)) throw new ArgumentNullException(instCode, "instCode cannot be null or empty");
```
Comment (timabell) [inline src/ui/Controllers/CourseController.cs]: Add braces.

https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/

Reject.

File: src/ui/Controllers/CourseController.cs (hunk 5 above)
Comment (MartinBentley) [reply src/ui/Controllers/CourseController.cs]: SingleOrDefault is now everywhere where only one is expected

File: src/ui/Controllers/CourseController.cs (hunk 6 above)
Comment (MartinBentley) [reply src/ui/Controllers/CourseController.cs]: All done

File: src/ui/Controllers/CourseController.cs (hunk 4 above)
Comment (MartinBentley) [reply src/ui/Controllers/CourseController.cs]: Good idea. Done

File: ManageCoursesUi.Tests/Helpers/TestHelper.cs (hunk 2 above)
Comment (MartinBentley) [reply ManageCoursesUi.Tests/Helpers/TestHelper.cs]: Done

**Pull request merged**